ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -42,54 +42,131 @@
   #include <unistd.h>
 #endif
 
+#ifndef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET
+#define HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET 0
+#endif
+
+#define RUBY_3_3_PLUS HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET
+
 // Used to mark function arguments that are deliberately left unused
 #ifdef __GNUC__
   #define UNUSED_ARG  __attribute__((unused))
 #else
   #define UNUSED_ARG
 #endif
 
+typedef struct {
+  VALUE thread;
```

Comment:
I just noticed this. If you store a VALUE reference in that struct, you MUST mark it. Otherwise it can be GCed and the object slot re-used, or GC compaction could move that object in another slot.

But if you do that, it also solves your question about how to get the thread name.

Reply:
> I just noticed this. If you store a VALUE reference in that struct, you MUST mark it. Otherwise it can be GCed and the object slot re-used, or GC compaction could move that object in another slot.

@byroot Thanks for the feedback here! Can you provide an example of how to do this in the context of this code? I've tried piecing it together hopping between gvltools and cruby, but hours in and I haven't gotten anything working. I can keep chipping away, but tbh the appropriate ways of when and where to mark in cruby are still pretty opaque to me. 

Same goes for things like `TypedData_Make_Struct`. I thought that might be a good approach because of the free/mark options it has, but I landed on this version because I just couldn't get that working.

> But if you do that, it also solves your question about how to get the thread name.

I think this lines up with @ivoanjo's comment: 

> (E.g. keep it in a C array, that we need to mark but otherwise can operate on without the GVL)

So i'm assuming a combo of mark + a large-ish C array (maybe 1024 like the CRuby instrumentation tests?) would suffice to hold onto the metadata I need so I can get the thread names once the GvlTracing block finishes.

I am curious - for something like [vernier](https://github.com/jhawthorn/vernier), which presumably would be used to wrap long running ruby processes, how would you expect it to efficiently keep track of thread names? For gvl-tracing, you're not terribly likely to be running 1k+ threads. But for something like vernier there really are no limits to what the code running inside of it might do.

Reply:
@jpcamara Not sure if you know this guide -- https://silverhammermba.github.io/emberb/c/#data -- it has a pretty good guide of how to setup a `TypedData` object, including the marking.

Currently, gvl-tracing is all class methods ("static"), but I guess to be able to mark we'd need to introduce a TypedData object, so maybe something like a `GvlTracing::State` object, which would contain inside it the threads array, that would then get marked.

> I am curious - for something like [vernier](https://github.com/jhawthorn/vernier), which presumably would be used to wrap long running ruby processes, how would you expect it to efficiently keep track of thread names? For gvl-tracing, you're not terribly likely to be running 1k+ threads. But for something like vernier there really are no limits to what the code running inside of it might do.

This is a good point. I haven't looked at how Vernier does it but we actually don't need to necessarily do it at the end of the tracing. We could check from time to time if a name is available (e.g. to avoid doing it on every event), and then render it and move on, rather than waiting until the end.

The "wait until the end" strategy was just a bit of a simplification ;)